p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,188 @@
+// over packetQueue and adds the ability to store addresses
+// from which this connection (indexed by ufrag) received
+// data.
+type muxedConnection struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	pq     *packetQueue
+	// list of remote addresses associated with this connection.
+	// this is useful as a mapping from [address] -> ufrag
+	addresses []string
+	ufrag     string
+	mux       *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	// TODO: double confirm that we really do not want to link this to a parent context
+	ctx, cancel := context.WithCancel(context.Background())
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) push(buf []byte, addr net.Addr) error {
+	return conn.pq.push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	// TODO: document why this is OK
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	// TODO: document why this is OK
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	// TODO: document why this is OK
+	return nil
+}
+
+// WriteTo implements net.PacketConn
+func (conn *muxedConnection) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return conn.mux.writeTo(p, addr)
+}
+
+func (conn *muxedConnection) closeConnection() error {
+	select {
+	case <-conn.ctx.Done():
+		return alreadyClosedErr
+	default:
+	}
+	conn.pq.close()
+	conn.cancel()
+	return nil
+}
+
+type packet struct {
+	addr net.Addr
+	buf  []byte
+}
+
+var (
+	errTooManyPackets = errors.New("too many packets in queue; dropping")
+)
+
+// just a convenience wrapper around a channel
+type packetQueue struct {
+	sync.Mutex
+	pkts   chan packet
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newPacketQueue() *packetQueue {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &packetQueue{
+		ctx:    ctx,
+		cancel: cancel,
+		pkts:   make(chan packet, maxPacketsInQueue),
+	}
+}
+
+// pop reads a packet from the packetQueue or blocks until
+// either a packet becomes available or the buffer is closed.
+func (pq *packetQueue) pop(ctx context.Context, buf []byte) (int, net.Addr, error) {
+	// TODO: see if this pattern with all this p and p.buf / pool business cannot
```

Comment:
For added context, the lifetime of a buffer is as follows:
1. Buffer is fetched from the global pool and passed to socket for reading.
2. If read is successful, mux then decides which connection to pass the buffer.
  - if no connection is found, the buffer is returned to the pool.
  - if pushing to the connection fails, the buffer is returned to the pool.
  - if pushing succeeds, the connection, and by extension the packet queue is considered as the buffer's owner.
3. Once the `pop` method is invoked, a buffer is dequeued, and it's contents are copied to the buffer provided in the method's argument. The dequeued buffer is then returned to the pool.

Reply:
Thanks, added it to the code for now, will see what to do with that documentation and code later.